airbyte-integrations/connectors/source-fastbill/source_fastbill/source.py

```diff
@@ -0,0 +1,320 @@
+        #     )
+
+
+class RecurringInvoices(FastbillStream):
+    primary_key = "INVOICE_ID"
+
+    def request_body_json(
+            self,
+            stream_state: Mapping[str, Any],
+            stream_slice: Mapping[str, Any] = None,
+            next_page_token: Mapping[str, Any] = None,
+    ) -> Optional[Union[Mapping, str]]:
+
+        if next_page_token:
+            return next_page_token
+        else:
+            return {
+                "SERVICE": "recurring.get",
+                "FILTER": {},
+                "OFFSET": 0
+            }
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        def req_body(offset):
+            return {
+                "SERVICE": "recurring.get",
+                "FILTER": {},
+                "OFFSET": offset
+            }
+
+        response = response.json()
+        offset = response["REQUEST"]["OFFSET"] if response["REQUEST"]["OFFSET"] >= 0 else None
+        if offset is None:
+            response_request = response["REQUEST"]["OFFSET"]
+            raise Exception(f"No valid offset value found:{response_request}")
+
+        if len(response["RESPONSE"]["INVOICES"]) == self.API_OFFSET_LIMIT:
+            return req_body(offset + self.API_OFFSET_LIMIT)
+        return None
+
+    def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapping]:
+        invoices = response.json().get("RESPONSE", {}).get("INVOICES", [])
+        yield from invoices
+        # for invoice in invoices:
+        #     yield self.schema_applier.apply_schema_transformations(
+        #         invoice,
+        #         self.get_json_schema()
+        #     )
+
+
+class Products(FastbillStream):
+    primary_key = "ARTICLE_ID"
+
+    def request_body_json(
+            self,
+            stream_state: Mapping[str, Any],
+            stream_slice: Mapping[str, Any] = None,
+            next_page_token: Mapping[str, Any] = None,
+    ) -> Optional[Union[Mapping, str]]:
+
+        if next_page_token:
+            return next_page_token
+        else:
+            return {
+                "SERVICE": "article.get",
+                "FILTER": {},
+                "OFFSET": 0
+            }
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        def req_body(offset):
+            return {
+                "SERVICE": "article.get",
+                "FILTER": {},
+                "OFFSET": offset
+            }
+
+        response = response.json()
+        offset = response["REQUEST"]["OFFSET"] if response["REQUEST"]["OFFSET"] >= 0 else None
+        if offset is None:
+            response_request = response["REQUEST"]["OFFSET"]
+            raise Exception(f"No valid offset value found:{response_request}")
+
+        if len(response["RESPONSE"]["ARTICLES"]) == self.API_OFFSET_LIMIT:
+            return req_body(offset + self.API_OFFSET_LIMIT)
+        return None
```

Comment:
You repeating a lot of code, please create function for those.

Reply:
Added a helper.py where I store these, I can move them back to the source if it' preferred.